posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,285 @@
+    extends Command
+
+  ...
+}
+----
+
+== Restoring the `Pinger` actor's protocol - Dotty version
+
+The Dotty version of the `Pinger` actor suffers from the same problem. The protocol of the actor is _widened_ with the `PingPong.Pong` message. This means that _any_ actor can send a `Pong` message to the `Pinger` actor. The solution in this case is different from previous one:
+
+* The `apply` method in the `Pinger` object, will return a `Behavior[Command]` thus restoring the original protocol
+* The `Pinger` actor defines an internal behavior of type `Behavior[CommandsAndResponses]`
+* The `Behavior[Command]` is derived from the `Behavior[CommandsAndResponses]` by applying the `narrow` method on the latter
+
+This leads to the following code:
+
+[source,scala,linenums]
+----
+object Pinger {
+
+  // My protocol
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+
+  // My protocol + the responses I need to understand...
+  private type CommandsAndResponses = Command | PingPong.Response
+
+  def apply(pingPong: ActorRef[PingPong.Ping]): Behavior[Command] = {
+    val internalBehavior: Behavior[CommandsAndResponses] =
+    Behaviors.setup { context =>
+      Behaviors.receiveMessage {
+        case StopPingPong =>
+          context.log.info(s"End of the ping-pong game - calling it a day!")
+          context.system.terminate()
+          Behaviors.stopped
+        case SendPing =>
+          pingPong ! PingPong.Ping(replyTo = context.self)
+          Behaviors.same
+        case response : PingPong.Response =>
+          context.log.info(s"Hey: I just received a $response !!!")
+          Behaviors.same
+      }
+    }
+    internalBehavior.narrow
+  }
+}
+----
+
+Admittedly, some magic has happened here and the question to ask is, _What is this `narrow` method doing_? Let's look at that in the next section.
+
+== Clearing up some magic
+
+The `Behavior` abstract class in Akka 2.6 defines the `narrow` method and here is a part of the relevant source code:
+
+[source,scala,linenums]
+----
+abstract class Behavior[T](private[akka] val _tag: Int) { behavior =>
+
+  /**
+   * Narrow the type of this Behavior, which is always a safe operation. This
+   * method is necessary to implement the contravariant nature of Behavior
+   * (which cannot be expressed directly due to type inference problems).
+   */
+  final def narrow[U <: T]: Behavior[U] = this.asInstanceOf[Behavior[U]]
+  
+  ...
+}
+  
+abstract class ExtensibleBehavior[T] extends Behavior[T](BehaviorTags.ExtensibleBehavior) {
+
+  ...
+  
+  def receive(ctx: TypedActorContext[T], msg: T): Behavior[T]
+  
+  ...
+----
+
+There's quite a bit going on here.
+
+First, we note that the `Behavior` class is generic: it has a type parameter `T`, which, because of no specific variance annotation on `T`, implies that `Behavior` is _invariant_ in its type parameter `T`. Also note the comment on the `narrow` method, stating: 
```

Comment:
Not necessary, but depending on the targeted audience, it might be nice to source where you get this info from that if there is no variance annotation, then it's invariant. I'm sure not everyone knows that, and they might be interested in reading why or where this can be found.

Reply:
https://docs.scala-lang.org/tour/variances.html

Reply:
@ckipp01 I just noticed that there still were some comments marked as not resolved. I actually link to the variance webpage earlier in the article. I'll add an inline link to the same doc.